File: labs/14-prompt-first-dev/go/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

const requirement = (
	"Build an API for a task manager. Users can create a task with a title and " +
		"optional description, list all tasks, and mark a task as complete. " +
		"Store tasks in memory only.")

func loadPromptTemplate(version string) (string, error) {
	promptPath := filepath.Join("prompts", version, "spec_gen.txt")
	if _, err := os.Stat(promptPath); err != nil {
		entries, _ := os.ReadDir("prompts")
		var available []string
		for _, e := range entries {
			if e.IsDir() {
				available = append(available, e.Name())
			}
		}
		return "", fmt.Errorf("prompt template not found for version '%s'. Available versions: %v", version, available)
	}
	b, err := os.ReadFile(promptPath)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func renderPrompt(template, requirement string) string {
	return strings.ReplaceAll(template, "{{ requirement }}", requirement)
}

func extractCodeBlocks(content string) (string, string, error) {
	mdRe := regexp.MustCompile("(?s)```markdown\\s*\\n(.*?)\\n```")
	pyRe := regexp.MustCompile("(?s)```python\\s*\\n(.*?)\\n```")
	mdMatch := mdRe.FindStringSubmatch(content)
	pyMatch := pyRe.FindStringSubmatch(content)
	if mdMatch == nil || pyMatch == nil {
		return "", "", fmt.Errorf("model response did not contain both required markdown and python code blocks")
	}
	return strings.TrimSpace(mdMatch[1]), strings.TrimSpace(pyMatch[1]), nil
}

func generateArtifacts(client *openaiclient.Client, requirement, version string) (string, string, error) {
	template, err := loadPromptTemplate(version)
	if err != nil {
		return "", "", err
	}
	prompt := renderPrompt(template, requirement)

	fmt.Printf("Using prompt version: %s\n", version)
	fmt.Printf("Sending %d prompt characters to model.\n", len(prompt))

	temp := 0.2
	maxTokens := 1500
	resp, err := client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages:    []openaiclient.Message{{Role: "user", Content: prompt}},
		Temperature: &temp,
		MaxTokens:   &maxTokens,
	})
	if err != nil {
		return "", "", err
	}
	content := resp.Choices[0].Message.Content
	return extractCodeBlocks(content)
}

func writeArtifacts(spec, scaffold string, outputDir string) error {
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return err
	}
	specPath := filepath.Join(outputDir, "api_spec.md")
	scaffoldPath := filepath.Join(outputDir, "scaffold.py")
	if err := os.WriteFile(specPath, []byte(spec+"\n"), 0644); err != nil {
		return err
	}
	if err := os.WriteFile(scaffoldPath, []byte(scaffold+"\n"), 0644); err != nil {
		return err
	}
	fmt.Printf("Wrote artifacts to %s\n", outputDir)
	fmt.Printf("  - %s\n", filepath.Base(specPath))
	fmt.Printf("  - %s\n", filepath.Base(scaffoldPath))
	return nil
}

func preview(text string, maxChars int) string {
	if len(text) <= maxChars {
		return text
	}
	return text[:maxChars] + "\n..."
}

func main() {
	client := openaiclient.NewClient()
	version := os.Getenv("PROMPT_VERSION")
	if version == "" {
		version = "v2"
	}

	spec, scaffold, err := generateArtifacts(client, requirement, version)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to generate artifacts: %v\n", err)
		os.Exit(1)
	}

	outputDir := "generated"
	if err := writeArtifacts(spec, scaffold, outputDir); err != nil {
		fmt.Fprintf(os.Stderr, "Failed to write artifacts: %v\n", err)
		os.Exit(1)
	}

	absOutput, _ := filepath.Abs(outputDir)
	fmt.Println("\n=== Generated API Spec Preview ===")
	fmt.Println(preview(spec, 400))
	fmt.Println("\n=== Generated Scaffold Preview ===")
	fmt.Println(preview(scaffold, 400))
	fmt.Printf("\nArtifacts written to: %s\n", absOutput)
}

File: labs/14-prompt-first-dev/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {

    private static final String REQUIREMENT = """
            Build an API for a task manager. Users can create a task with a title and \
            optional description, list all tasks, and mark a task as complete. \
            Store tasks in memory only.
            """;

    private static final String PROMPT_VERSION = System.getenv().getOrDefault("PROMPT_VERSION", "v2");

    private static final Pattern MARKDOWN_BLOCK = Pattern.compile(
            "```markdown\\s*\\n(.*?)\\n```", Pattern.DOTALL);
    private static final Pattern PYTHON_BLOCK = Pattern.compile(
            "```python\\s*\\n(.*?)\\n```", Pattern.DOTALL);

    public static void main(String[] args) throws Exception {
        OpenAiClient client = new OpenAiClient();
        ObjectMapper mapper = new ObjectMapper();

        String template = loadPromptTemplate(PROMPT_VERSION);
        String prompt = template.replace("{{ requirement }}", REQUIREMENT);

        System.out.println("Using prompt version: " + PROMPT_VERSION);
        System.out.println("Sending " + prompt.length() + " prompt characters to model.");

        ObjectNode payload = mapper.createObjectNode();
        ArrayNode messages = mapper.createArrayNode();
        ObjectNode user = messages.addObject();
        user.put("role", "user");
        user.put("content", prompt);
        payload.set("messages", messages);
        payload.put("temperature", 0.2);
        payload.put("max_tokens", 1500);

        JsonNode response = client.chatCompletion(payload);
        String content = client.extractMessage(response).get("content").asText();

        Map.Entry<String, String> artifacts = extractCodeBlocks(content);
        String spec = artifacts.getKey();
        String scaffold = artifacts.getValue();

        Path outputDir = Paths.get("..", "generated");
        writeArtifacts(spec, scaffold, outputDir);

        System.out.println("\n=== Generated API Spec Preview ===");
        System.out.println(preview(spec));
        System.out.println("\n=== Generated Scaffold Preview ===");
        System.out.println(preview(scaffold));
        System.out.println("\nArtifacts written to: " + outputDir.toAbsolutePath());
    }

    private static String loadPromptTemplate(String version) throws Exception {
        Path promptPath = Paths.get("..", "python", "prompts", version, "spec_gen.txt").toAbsolutePath().normalize();
        if (!Files.exists(promptPath)) {
            throw new IllegalArgumentException("Prompt template not found for version '" + version + "' at " + promptPath);
        }
        return Files.readString(promptPath);
    }

    private static Map.Entry<String, String> extractCodeBlocks(String content) {
        Matcher mdMatch = MARKDOWN_BLOCK.matcher(content);
        Matcher pyMatch = PYTHON_BLOCK.matcher(content);
        if (!mdMatch.find() || !pyMatch.find()) {
            throw new IllegalArgumentException("Model response did not contain both required markdown and python code blocks.");
        }
        return Map.entry(mdMatch.group(1).trim(), pyMatch.group(1).trim());
    }

    private static void writeArtifacts(String spec, String scaffold, Path outputDir) throws Exception {
        Files.createDirectories(outputDir);
        Path specPath = outputDir.resolve("api_spec.md");
        Path scaffoldPath = outputDir.resolve("scaffold.py");
        Files.writeString(specPath, spec + "\n");
        Files.writeString(scaffoldPath, scaffold + "\n");
        System.out.println("Wrote artifacts to " + outputDir);
        System.out.println("  - " + specPath.getFileName());
        System.out.println("  - " + scaffoldPath.getFileName());
    }

    private static String preview(String text) {
        int maxChars = 400;
        if (text.length() <= maxChars) {
            return text;
        }
        return text.substring(0, maxChars).stripTrailing() + "\n...";
    }
}
